text/0000-nested-publish.md

```diff
@@ -0,0 +1,186 @@
+
+* Clutters the public view of the registry with packages not intended to be usable on their own, and which may even become obsolete as internal architecture changes.
+* Requires multiple `cargo publish` operations (this could be fixed with bulk publication) and writing public metadata for each package.
+* Can result in semver violations and thus compilation failures, due to the developer not thinking about semver compatibility within the group.
+
+This RFC will allow developers to avoid all of these inconveniences and hazards by publishing a single package.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default (and always, prior to this RFC's implementation):
+
+* If your package contains any sub-packages, Cargo [excludes](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) them from the `.crate` archive file produced by `cargo package` and `cargo publish`.
+* If your package contains any non-`dev` dependencies which do not give a `version = "..."`, it cannot be published to `crates.io`.
+* If your package contains `[dev-dependencies]` which do not give a `version = "..."`, they are stripped out on publication.
+
+(By “**sub-package**” we mean a package (directory with `Cargo.toml`) which is a subdirectory of another package. We shall call the outermost such package, the package being published, the “**parent package**”.)
+
+You can change this default by placing in the manifest (`Cargo.toml`) of a sub-package:
+
+```toml
+[package]
+publish = "nested"
+```
+
+If this is done, Cargo's behavior changes as follows:
+
+* If you publish the parent package, the sub-package is included in the `.crate` file (unless overridden by explicit `exclude`/`include`) and will be available to the parent package whenever the parent package is downloaded and compiled.
+* The parent package (and other sub-packages) may have `path =` dependencies upon the sub-package. (Such dependencies must not have a `version =` or `git =`; that is, the `path` must be the _only_ source for the dependency.)
+* You cannot `cargo publish` the sub-package, just as if it had `publish = false`. (This is a safety measure against accidentally publishing the sub-package separately when this is not intended.)
+
+Nested sub-packages may be freely placed within other nested sub-packages.
+
+When a group of packages is published in this way, and depended on, this has a number of useful effects (which are not things that Cargo explicitly implements, just consequences of the system):
+
+* The packages are a single unit for all versioning purposes; there is no way for a version mismatch to arise since all the code was published together. Version resolution does not apply (in the same way that it does not for any other `path =` dependency).
+* The sub-package is effectively “private”: it cannot be named by any other package on `crates.io`, only by its parent package and sibling sub-packages.
```

Comment:
This should probably be clarified that it's private insomuch as it's not automatically public. It would still be possible to re-export something (or everything).

Reply:
See also https://github.com/rust-lang/rfcs/pull/3452#discussion_r1251086543

Reply:
With "effectively", I feel like this is trying to describe the semantics and then describe the relevant mechanism.

For me, I like the first part (effectively "private").  For the second, I'd change it to "it cannot be named **as a dependency** by any other package...".  I think this would clarify that "private" is referring to the package aspect and not what is being done within the Rust APIs.

Reply:
I have added “cannot be named as a dependency” and explicitly mentioned that re-exports are still possible.